File: tests/support.py
import importlib.util
import json
import os
import subprocess
import tempfile
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]

# Fixture git repos must not inherit the developer's global/system git config —
# commit signing, hook paths, and templates break `git commit` on machines
# where signing is interactive, while CI runners stay green.
GIT_ENVIRONMENT = dict(
    os.environ,
    GIT_CONFIG_GLOBAL=os.devnull,
    GIT_CONFIG_SYSTEM=os.devnull,
)


def load_module(name, relative_path):
    path = ROOT / relative_path
    spec = importlib.util.spec_from_file_location(name, path)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def fixture_directory(case):
    """A per-test TemporaryDirectory cleaned via addCleanup, so it is removed
    deterministically even when setUp fails partway."""
    temporary_directory = tempfile.TemporaryDirectory()
    case.addCleanup(temporary_directory.cleanup)
    return Path(temporary_directory.name)


def write(path, content):
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content, encoding="utf-8")


def write_json(path, value):
    write(path, json.dumps(value, indent=2) + "\n")


def write_plugin_manifest(root, name, **fields):
    write_json(
        root / f"plugins/{name}/.claude-plugin/plugin.json",
        {"name": name, **fields},
    )


def write_plugin_registry(root, vendored_skills, dual_plugins, claude_only=None):
    write_json(
        root / "plugin-registry.json",
        {
            "vendored_skills": vendored_skills,
            "dual_harness_plugins": dual_plugins,
            "claude_only_plugins": claude_only or {},
        },
    )


def run(command, cwd, check=True, env=None):
    return subprocess.run(
        command,
        cwd=cwd,
        capture_output=True,
        text=True,
        check=check,
        env=env,
    )


def git(repo, *args, check=True):
    return run(["git", *args], repo, check=check, env=GIT_ENVIRONMENT)


def initialize_git_repo(repo):
    git(repo, "init", "-b", "main")
    git(repo, "config", "user.name", "Test User")
    git(repo, "config", "user.email", "test@example.com")


def commit_all(repo, message="fixture"):
    git(repo, "add", "-A")
    git(repo, "commit", "-m", message)

File: tooling/test/support.mjs
import { spawnSync } from "node:child_process";
import { mkdirSync, mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { tmpdir, devNull } from "node:os";

const gitEnvironment = {
  ...process.env,
  GIT_CONFIG_GLOBAL: devNull,
  GIT_CONFIG_SYSTEM: devNull,
};

export function temporaryDirectory(testContext) {
  const path = mkdtempSync(join(tmpdir(), "plugin-sync-test-"));
  testContext.after(() => rmSync(path, { recursive: true, force: true }));
  return path;
}

export function writeText(path, content) {
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, content, "utf8");
}

export function writeJson(path, value) {
  writeText(path, `${JSON.stringify(value, null, 2)}\n`);
}

export function writePluginManifest(root, name, fields = {}) {
  writeJson(join(root, `plugins/${name}/.claude-plugin/plugin.json`), {
    name,
    ...fields,
  });
}

export function writePluginRegistry(root, vendoredSkills, dualPlugins, claudeOnly = {}) {
  writeJson(join(root, "scripts/plugin-registry.json"), {
    vendored_skills: vendoredSkills,
    dual_harness_plugins: dualPlugins,
    claude_only_plugins: claudeOnly,
  });
}

export function run(command, arguments_, cwd, options = {}) {
  const result = spawnSync(command, arguments_, {
    cwd,
    encoding: "utf8",
    env: options.git ? gitEnvironment : process.env,
  });
  if (options.check !== false && result.status !== 0) {
    throw new Error(
      `${command} ${arguments_.join(" ")} failed (${String(result.status)}):\n${result.stdout}\n${result.stderr}`,
    );
  }
  return result;
}

export function git(root, ...arguments_) {
  return run("git", arguments_, root, { git: true });
}

export function initializeGitRepository(root) {
  git(root, "init", "-b", "main");
  git(root, "config", "user.name", "Test User");
  git(root, "config", "user.email", "test@example.com");
}

export function commitAll(root, message = "fixture") {
  git(root, "add", "-A");
  git(root, "commit", "-m", message);
}
